edx-solutions/edx-platform
# [BB-2959] Mcka User progress migration endpoint and background task

This PR adds endpoint and background task, that can be used to migrate course progress of some user (enrollment, block completions, module states, submissions). 

**Testing instructions**:

1. Run following script. Follow comments.
    ```python
    #!/usr/bin/python3

    import csv
    import requests

    from io import StringIO


    ### Prepare csv with data of users to be migrated


    # You have to create course and two users.
    # Make sure you have toggled ENABLE_COMPLETION_TRACKING waffle switch. You can do this in admin panel.
    # After that, create some blocks in course, enroll source user in this course and submit answers for blocks.
    row = ('course-v1:OpenCraft+Course1+C1', 'source@example.com', 'target@example.com', '')

    file = StringIO()
    writer = csv.writer(file)
    csv_header = ('course', 'source_email', 'dest_email', 'outcome')
    writer.writerow(csv_header)
    writer.writerow(row)

    file.seek(0)

    ### Send request to migrate user progress

    migrate_url = 'http://edx.devstack.lms:18000/api/user/v1/progress/migrate/'

    # Relogin in LMS, make any request to LMS, then open browser debugger, and retreive these
    # values from request's `Cookie` header
    jwt_header = 'eyJhbGciOiJSUzUxMiIsImtpZCI6ImRldnN0YWNrX2tleSJ9'
    jwt_payload = 'eyJzY29wZXMiOiBbImVtYWlsIiwgInByb2ZpbGUiXSwgImFkbWluaXN0cmF0b3IiOiB0cnVlLCAic3VwZXJ1c2VyIjogdHJ1ZSwgInByZWZlcnJlZF91c2VybmFtZSI6ICJlZHgiLCAiZmlsdGVycyI6IFsidXNlcjptZSJdLCAiaWF0IjogMTYwMzk5MTAzMywgInN1YiI6ICI1MGNmOTJmOWEzZDg0ODlkZjk1ZTU4MzU0OWI5MTlkZiIsICJmYW1pbHlfbmFtZSI6ICIiLCAiYXVkIjogImxtcy1rZXkiLCAiaXNzIjogImh0dHA6Ly9sb2NhbGhvc3Q6MTgwMDAvb2F1dGgyIiwgImVtYWlsX3ZlcmlmaWVkIjogdHJ1ZSwgIm5hbWUiOiAiIiwgInZlcnNpb24iOiAiMS4xLjAiLCAiZ2l2ZW5fbmFtZSI6ICIiLCAiZXhwIjogMTYwMzk5NDYzMywgImlzX3Jlc3RyaWN0ZWQiOiBmYWxzZSwgImVtYWlsIjogImVkeEBleGFtcGxlLmNvbSJ9'
    jwt_signature = 'G1fxJFH1zPZjs_ZT1e_oj_qpE3-9Wb0N4QN_RooRlXVVNHuYnUUBVi5p6W_QAJdv2V2EJiwzaDk4nGuOo1iV0-p0aCawg9XASePxWdskyyW74elyX94rmEeLn4-u2wjw7GxOx8bcZk_vXPeERoiMyFdVuXoJL3O8H-gQmkxe81sG8iJ93QZ4l-4D4fPJltL5O_UT7POyjBPQG-TEsjIQd8wcn4zdcDdwjh3Td3HzBuiKeVc-U9_rJrtQcmi8uZ-eQJmeq0u7gI-Lg9H-3X_of_cXe8A7xnB0lRetx9I6DDoHdrnP83bkZ5PUnhSZo-u3zP1G54yGcCSzHUiw64pApA'

    jwt_token = '{}.{}.{}'.format(jwt_header, jwt_payload, jwt_signature)

    headers = {
        'Authorization': 'JWT {}'.format(jwt_token)
    }

    requests.post(migrate_url, files={'file': file}, headers=headers)
    ```
2. You should see something like this in console of your devstack:
    ```
    2020-10-29 17:35:35,157 INFO 6099 [openedx.core.djangoapps.user_api.completion.tasks] tasks.py:42 - Started progress migration. Items to process: 1
    2020-10-29 17:35:35,157 INFO 6099 [openedx.core.djangoapps.user_api.completion.tasks] tasks.py:99 - Started progress migration from "source@example.com" to "target1@example.com" for "course-v1:OpenCraft+Course1+C1" course
    2020-10-29 17:35:35,211 INFO 6099 [completion_aggregator.batch] batch.py:104 - Performing aggregation update for 1 user enrollments
    2020-10-29 17:35:35,212 INFO 6099 [completion_aggregator.tasks.aggregation_tasks] aggregation_tasks.py:62 - Updating aggregators in course-v1:OpenCraft+Course1+C1 for target1. Changed blocks: set([BlockUsageLocator(CourseLocator(u'OpenCraft', u'Course1', u'C1', None, None), u'about', u'short_description')])
    2020-10-29 17:35:35,232 INFO 6099 [completion_aggregator.batch] batch.py:126 - Finished aggregation update for 1 user enrollments
    2020-10-29 17:35:35,252 INFO 6099 [completion_aggregator.batch] batch.py:104 - Performing aggregation update for 1 user enrollments
    2020-10-29 17:35:35,253 INFO 6099 [completion_aggregator.tasks.aggregation_tasks] aggregation_tasks.py:62 - Updating aggregators in course-v1:OpenCraft+Course1+C1 for target1. Changed blocks: set([BlockUsageLocator(CourseLocator(u'OpenCraft', u'Course1', u'C1', None, None), u'about', u'overview')])
    2020-10-29 17:35:35,270 INFO 6099 [completion_aggregator.batch] batch.py:126 - Finished aggregation update for 1 user enrollments
    2020-10-29 17:35:35,272 INFO 6099 [openedx.core.djangoapps.user_api.completion.tasks] tasks.py:178 - User progress in "course-v1:OpenCraft+Course1+C1" course successfully migrated from "source@example.com" to "target1@example.com"
    2020-10-29 17:35:35,273 INFO 6099 [openedx.core.djangoapps.user_api.completion.tasks] tasks.py:93 - Email with users progress migration results sent to [u'devops@example.com']
    ```
3. You should see target user enrolled to the course and with the same submissions as first user.


**Author notes and concerns**: None

**Reviewers**
